lib/logmatcher.c

```diff
@@ -355,20 +375,23 @@ log_matcher_pcre_re_compile(LogMatcher *s, const gchar *re, GError **error)
       return FALSE;
     }
 
+  _alloc_matches(self, re);
+
   return TRUE;
 }
 
 static void
-log_matcher_pcre_re_feed_backrefs(LogMatcher *s, LogMessage *msg, gint value_handle, int *matches, gint match_num,
+log_matcher_pcre_re_feed_backrefs(LogMatcher *s, LogMessage *msg, gint value_handle, gint match_num,
                                   const gchar *value)
 {
   gint i;
   gboolean indirect = _shall_set_values_indirectly(value_handle);
+  LogMatcherPcreRe *self = (LogMatcherPcreRe *) s;
 
   for (i = 0; i < (RE_MAX_MATCHES) && i < match_num; i++)
     {
-      gint begin_index = matches[2 * i];
-      gint end_index = matches[2 * i + 1];
+      gint begin_index = self->matches[2 * i];
```

Comment:
I'm afraid matches was not part of the state of `LogMatchrePcreRe` for a reason. The `log_rewrite_subst_process` (and every filter/parser/rewrite in general) runs in source threads, but they are using the same class instance. This change introduces a race condition: the different source threads are using the same `self->matches` buffer.

Reply:
Agreed! With the new patch, I only moved the `pcre_fullinfo()` call to `pcre_re_compile()`. Thanks!